xcube/webapi/ows/coverages/scaling.py

```diff
@@ -0,0 +1,135 @@
+
+        self._crs = crs
+        self._y_name = self.get_axis_from_crs(
+            {'lat', 'latitude', 'geodetic latitude', 'n', 'north', 'y'}
+        )
+        self._x_name = self.get_axis_from_crs(
+            {'longitude', 'geodetic longitude', 'lon', 'e', 'east', 'x'}
+        )
+
+        # The standard doesn't define behaviour when multiple scale
+        # parameters are given. We choose to handle one and ignore the
+        # others in such cases.
+        if request.scale_factor is not None:
+            self._scale = request.scale_factor, request.scale_factor
+        elif request.scale_axes is not None:
+            self._scale = self._get_xy_values(request.scale_axes)
+        elif request.scale_size is not None:
+            # The standard allows floats for "scale-size" but mandates:
+            # "The returned coverage SHALL contain exactly the specified number
+            # of sample values". We can't return a fractional number of sample
+            # values, so truncate to int here.
+            x, y = self._get_xy_values(request.scale_size)
+            self._final_size = int(x), int(y)
+        else:
+            # The standard doesn't mandate this as a default; in future, we
+            # might choose to downscale automatically if a large coverage
+            # is requested without an explicit scaling parameter.
+            self._scale = (1, 1)
+
+    @property
+    def scale(self) -> tuple[float, float]:
+        if self._scale is not None:
+            return self._scale
+        else:
+            x_initial, y_initial = self._initial_size
+            x_final, y_final = self._final_size
+            return x_initial / x_final, y_initial / y_final
+
+    @property
+    def size(self) -> tuple[float, float]:
+        if self._final_size is not None:
+            return self._final_size
+        else:
+            x_initial, y_initial = self._initial_size
+            x_scale, y_scale = self._scale
+            return x_initial / x_scale, y_initial / y_scale
+
+    def _get_xy_values(
+        self, axis_to_value: dict[str, float]
+    ) -> tuple[float, float]:
+        x, y = None, None
+        for axis in axis_to_value:
+            if axis.lower()[:3] in ['x', 'e', 'eas', 'lon', self._x_name]:
+                x = axis_to_value[axis]
+            if axis.lower()[:3] in ['y', 'n', 'nor', 'lat', self._y_name]:
+                y = axis_to_value[axis]
+        return x, y
+
+    def get_axis_from_crs(self, valid_identifiers: set[str]):
+        for axis in self._crs.axis_info:
+            if not hasattr(axis, 'abbrev'):
+                continue
+            identifiers = set(
+                map(
+                    lambda attr: getattr(axis, attr, '').lower(),
+                    ['name', 'abbrev'],
+                )
+            )
+            if identifiers & valid_identifiers:
+                return axis.abbrev
+        return None
+
+    def apply(self, gm: GridMapping):
```

Comment:
```suggestion
    def apply(self, gm: GridMapping) -> GridMapping:
```

Also, a tiny bit of documentation would be nice

Reply:
Type hint added, along with docstrings and comments for the whole class -- the implementation is simple but has one or two subtleties that are worth spelling out, so thanks for the reminder to document.